### PR TITLE
Dev worker cdk

### PR DESCRIPTION
**Pull request by alexiswl**
Last merge didn't add these final commits (including 9eea9e525f10b674402d17996ffd0215b13ed970)

### CODE REVIEW

File: cdk/apps/dev_worker/README.md [hunk 1]
```diff
@@ -60,7 +66,7 @@ Context parameters are defined with the `-c "STACK_NAME=my-stack` argument.
 Repeat the `-c` parameter for defining multiple parameters.    
 i.e 
 ```
-cdk deploy -c "STACK_NAME=my-stack"
+cdk deploy -c "STACK_NAME=my-stack" -c "USE_SPOT_INSTANCE=false"
```
Comment (brainstorm) [inline cdk/apps/dev_worker/README.md:69]: @alexiswl Why? SPOT instances are **very** much preferred, by default :)

File: cdk/apps/dev_worker/README.md [hunk 2]
```diff
@@ -101,18 +108,45 @@ cdk destroy -c "STACK_NAME=alexis-unique-stack"
...
 ```
 host i-* mi-*
     ProxyCommand sh -c "aws ssm start-session --target %h --document-name AWS-StartSSHSession --parameters 'portNumber=%p'"
     User ec2-user
     IdentityFile /path/to/private/key/.ssh/aws
 ```
 
+### Setting port forwarding through SSM
+
+Set up the following shell function in your `~.bashrc`.
```
Comment (brainstorm) [inline cdk/apps/dev_worker/README.md:123]: @alexiswl Fantastic! <3

File: cdk/apps/dev_worker/user_data/user_data.sh [hunk 3]
```diff
@@ -128,4 +128,12 @@ rm Anaconda3-${!ANACONDA_VERSION}-Linux-x86_64.sh
 
 # Fix bashrc for ec2-user and ssm-user for access ready for conda
 su - "ec2-user" -c "/opt/conda/bin/conda init"
-su - "ssm-user" -c "/opt/conda/bin/conda init"
\ No newline at end of file
+su - "ssm-user" -c "/opt/conda/bin/conda init"
+
+# Pull starter notebook into home directory
+# FIXME update notebook path once in master branch
+starter_notebook="https://raw.githubusercontent.com/alexiswl/umccr-infrastructure/dev_worker_cdk/cdk/apps/dev_worker/dev_worker_notebook.ipynb"
```
Comment (brainstorm) [inline cdk/apps/dev_worker/user_data/user_data.sh:135]: Brilliant! :)

File: cdk/apps/dev_worker/README.md (hunk 1 above)
Comment (reisingerf) [reply cdk/apps/dev_worker/README.md:69]: Happy for it to be configurable, as long as the default is SPOT.
However, I would remove the spot price parameter. It's kind of annoying to first find out what the current spot price is and then adjust it each time you want to change the instance type.

**Comment by alexiswl** [comment]
So the m5 doesn't work just yet, my guess is there's problems with the eb2 volumes mount. Let me re-look into this.

**Comment by brainstorm** [comment]
Weird, it did work when I tested it myself :-S What's failing on your end?

**Comment by alexiswl** [comment]
> Weird, it did work when I tested it myself :-S What's failing on your end?

Not sure, and working fine through a new ami type - the current one is about ten months old. It could have actually been from not running cdk destroy beforehand and having left over stack problems. Few other teething issues I've noticed with the userdata that I'll need to fix up before merging.

File: cdk/apps/dev_worker/README.md (hunk 1 above)
Comment (alexiswl) [reply cdk/apps/dev_worker/README.md:69]: Can't pause a spot instance and every now and again they go away which can be annoying. I think for the dev worker use case there's definitely a need for non-spot instances.

Comment (alexiswl) [reply cdk/apps/dev_worker/README.md:69]: A default spot price parameter needs to be set though right?

Comment (reisingerf) [reply cdk/apps/dev_worker/README.md:69]: You don't need a spot price parameter (IIRC), only setting the `MarketType` to `spot` should use the default spot price (which is usually the on-demand price).

Comment (alexiswl) [reply cdk/apps/dev_worker/README.md:69]: So one should never need to set the spot price? Or should this remain as an option?

**Comment by alexiswl** [comment]
Also I don't like keeping the "STACK_NAME" inside the cdk.json file, it syncs with git by default and I find myself performing continuous soft resets. Thoughts on having a file called .stack_name which is read in / written by cdk? And we can add this to .gitignore?

File: cdk/apps/dev_worker/README.md (hunk 1 above)
Comment (reisingerf) [reply cdk/apps/dev_worker/README.md:69]: I would vote for the former. The less effort the better and setting a spot price explicitly is only useful in very rare cases.

**Comment by reisingerf** [comment]
TBH I would keep it as simple as possible and keep a user specific default stack name, as previously suggested. 
There should not be any need for cdk.json hacks and it should work for the majority of use cases.

**Comment by alexiswl** [comment]
> TBH I would keep it as simple as possible and keep a user specific default stack name, as previously suggested.
> There should not be any need for cdk.json hacks and it should work for the majority of use cases.

I've amended this as so.

1. A file called .stack_name (which is part of git ignore) is created in the case that a uuid stack is used.

The hierachy is as follows.
Context (-c) -> .stack_name -> new uuid.

If a stack name is provided as part of the context parameter, the .stack_name is not updated. To refresh a new uuid stack_name, the .stack_name file can be deleted and a new one will created on the next cdk command.

**Comment by brainstorm** [comment]
@alexiswl Will that `.stack_name` will be kept on the current repository? $HOME? Elsewhere?

As long as it doesn't produce confusion among users I'm fine as it is ;)

**Comment by alexiswl** [comment]
Just in the same directory as cdk.json and cdk.context.json. I added the file to the .gitignore too.

**Comment by brainstorm** [comment]
Alrighty, shall we merge as-is and let @pdiakumis give it a go (he volunteered as pro end-user)? ;)

**Pull request merged**